File: src/main/java/no/nav/nare/core/regelsettyper/Modrekvote.java
package no.nav.nare.core.regelsettyper;

import no.nav.nare.core.specifications.Specification;

import static no.nav.nare.input.Rolle.FAR;
import static no.nav.nare.input.Rolle.MOR;
import static no.nav.nare.input.Soknadstype.ADOPSJON;
import static no.nav.nare.input.Soknadstype.FODSEL;
import static no.nav.nare.input.Uttaksplan.INNEN_3_AAR;
import static no.nav.nare.input.Uttaksplan.SAMMENHENGENDE;

import static no.nav.nare.core.regelsettyper.regler.HarRettTilForeldrePenger.harRettTilForeldrePenger;
import static no.nav.nare.core.regelsettyper.regler.HarUttaksplanForModreKvote.harUttaksplanForModreKvoteAdopsjonl;
import static no.nav.nare.core.regelsettyper.regler.HarUttaksplanForModreKvote.harUttaksplanForModreKvoteFodsel;
import static no.nav.nare.core.regelsettyper.regler.SoknadGjelder.søknadGjelder;
import static no.nav.nare.core.specifications.NotSpecification.ikke;

public class Modrekvote extends Ruleset {


    public Modrekvote() {
        specification = getModreKvote();
    }

    public Specification getModreKvote() {

        Specification harBeggeForeldreRettTilForeldrepenger =
                regel("FK_VK_10.1", "Har begge foreldre rett til foreldrepenger?", harRettTilForeldrePenger(MOR).og(harRettTilForeldrePenger(FAR)));

        Specification gjelderSøknadFødsel =
                regel("FK_VK 10.2", "Gjelder søknad fødsel?", søknadGjelder(FODSEL));

        Specification gjelderSøknadAdopsjon =
                regel("FK_VK 10.3", "Gjelder søknad adopsjon?", søknadGjelder(ADOPSJON));

        Specification harUttaksplanEtterFodsel =
                regel("FK_VK_10.4", "Har mor uttaksplan sammenhengende eller tre år etter fødsel?", harUttaksplanForModreKvoteFodsel(SAMMENHENGENDE).eller(harUttaksplanForModreKvoteFodsel(INNEN_3_AAR)));

        Specification harUttaksplanEtterAdopsjon =
                regel("FK_VK_10.5", "Har mor uttaksplan sammenhengende eller tre år etter adopsjon?", harUttaksplanForModreKvoteAdopsjonl(INNEN_3_AAR));

        Specification vilkårForFødsel =
                regel("FK_VK.10.A", harBeggeForeldreRettTilForeldrepenger.og(gjelderSøknadFødsel).og(harUttaksplanEtterFodsel));

        Specification vilkårForAdopsjon =
                regel("FK_VK.10.B", harBeggeForeldreRettTilForeldrepenger
                        .og(ikke(gjelderSøknadFødsel).medBeskrivelse("søknad gjelder ikke fødsel"))
                        .og(gjelderSøknadAdopsjon)
                        .og(harUttaksplanEtterAdopsjon));

        return regel("FK_VK.10", "Er vilkår for mødrekvote oppfylt for enten fødsel eller adopsjon?", vilkårForFødsel.eller(vilkårForAdopsjon));
    }


}

File: src/test/kotlin/no/nav/nare/demo/Modrekvote.kt
package no.nav.nare.demo

import no.nav.nare.core.demo.Rolle
import no.nav.nare.core.demo.Rolle.FAR
import no.nav.nare.core.demo.Rolle.MOR
import no.nav.nare.core.demo.Soknad
import no.nav.nare.core.demo.Soknadstype
import no.nav.nare.core.demo.Soknadstype.ADOPSJON
import no.nav.nare.core.demo.Soknadstype.FODSEL
import no.nav.nare.core.demo.Uttaksplan.INNEN_3_AAR
import no.nav.nare.core.demo.Uttaksplan.SAMMENHENGENDE
import no.nav.nare.core.demo.Uttaksplan.SENERE
import no.nav.nare.core.evaluations.Evaluering
import no.nav.nare.core.evaluations.Evaluering.Companion.ja
import no.nav.nare.core.evaluations.Evaluering.Companion.kanskje
import no.nav.nare.core.evaluations.Evaluering.Companion.nei
import no.nav.nare.core.specifications.Spesifikasjon

class Regelsett {

   private val morHarRettTilForeldrepenger = Spesifikasjon<Soknad>(
      identifikator = "FK_VK_10.1",
      beskrivelse = "Har søker med rolle mor rett til foreldrepenger?",
      implementasjon = { søkerHarPåkrevdRolle(MOR, this) }
   )

   private val farHarRettTilForeldrepenger = Spesifikasjon<Soknad>(
      identifikator = "FK_VK_10.1",
      beskrivelse = "Har søker med rolle fae rett til foreldrepenger?",
      implementasjon = { søkerHarPåkrevdRolle(FAR, this) }
   )

   private val gjelderSoknadFødsel = Spesifikasjon<Soknad>(
      identifikator = "FK_VK 10.2",
      beskrivelse = "Soknad gjelder fødsel?",
      implementasjon = { soknadGjelder(FODSEL, this) }
   )

   private val gjelderSoknadAdopsjon = Spesifikasjon<Soknad>(
      identifikator = "FK_VK 10.23",
      beskrivelse = "Soknad gjelder adopsjon?",
      implementasjon = { soknadGjelder(ADOPSJON, this) }
   )


   private val harUttaksplanEtterFodsel = Spesifikasjon<Soknad>(
      identifikator = "FK_VK 10.4/FK_VK 10.5",
      beskrivelse = "Foreligger det korrekt uttaksplan etter fødsel?",
      implementasjon = { harUttaksplanForModreKvote(FODSEL, this)
      }
   )

   private val harUttaksplanEtterAdopsjon = Spesifikasjon<Soknad>(
      identifikator = "FK_VK 10.6",
      beskrivelse = "Foreligger det korrekt uttaksplan etter adopsjon?",
      implementasjon = { harUttaksplanForModreKvote(ADOPSJON, this) }
   )

   private val harBeggeForeldreRettTilForeldrepenger = (farHarRettTilForeldrepenger og morHarRettTilForeldrepenger).med(
      identifikator = "FK_VK_1.3",
      beskrivelse = "Har begge foreldre rett til foreldrepenger?"
   )
   private val vilkårForFødsel =(harBeggeForeldreRettTilForeldrepenger og gjelderSoknadFødsel og harUttaksplanEtterFodsel).med(
      identifikator = "FK_VK 1.11",
      beskrivelse = "Er vilkår for fødsel oppfylt"
   )

   private val vilkårForAdopsjon =(
      harBeggeForeldreRettTilForeldrepenger og
         gjelderSoknadFødsel.ikke() og
         gjelderSoknadAdopsjon og
         harUttaksplanEtterAdopsjon).med(
      identifikator = "FK_VK 1.2",
      beskrivelse = "Er vilkår for adopsjon oppfylt?"
   )

   val mødrekvote = (vilkårForFødsel eller vilkårForAdopsjon).med(
      identifikator = "FK_VK_1",
      beskrivelse =  "Er vilkår for mødrekvote oppfylt?"
   )

}

fun soknadGjelder(søknadstype: Soknadstype, søknad: Soknad): Evaluering =
   if (søknad.søknadstype == søknadstype)
      ja("Søknad gjelder fødsel")
   else
      nei("Søknad gjelder ikke fødsel")

fun søkerHarPåkrevdRolle(rolle: Rolle, søknad: Soknad): Evaluering =
      søknad.hentSøkerIRolle(rolle)?.let { person ->
      if (person.rettTilFp)
         ja("Søker med rolle $rolle har rett til foreldrepenger")
      else
         nei("Søker med rolle $rolle har ikke rett til foreldrepenger")
   } ?: nei("Søker har ikke rolle $rolle")


fun harUttaksplanForModreKvote(soknadstype: Soknadstype, søknad: Soknad): Evaluering =
   søknad.hentSøkerIRolle(MOR)?.let { mor ->
      mor.uttaksplan?.let { morsUttaksplan ->
         when(morsUttaksplan){
            SAMMENHENGENDE -> ja("Mødrekvote tas sammenhengende etter $soknadstype")
            INNEN_3_AAR -> nei("Mødrekvote tas ikke senere enn 3 år etter $soknadstype")
            SENERE -> kanskje("Saksbehandler må se på dette")
         }
      } ?: nei("Det foreligger ingen uttaksplan for mor")
   } ?: nei("Søker er ikke mor")





